numba/ccallback.py

```diff
@@ -60,16 +60,17 @@ def enable_caching(self):
 
     def compile(self):
         # Try to load from cache
-        cres = self._cache.load_overload(self._sig, self._targetdescr.target_context)
-        if cres is None:
-            cres = self._compile_uncached()
-            self._cache.save_overload(self._sig, cres)
-        else:
-            self._cache_hits += 1
-
-        self._library = cres.library
-        self._wrapper_name = cres.fndesc.llvm_cfunc_wrapper_name
-        self._wrapper_address = self._library.get_pointer_to_function(self._wrapper_name)
+        with compiler.lock_compiler:
```

Comment:
This puts all of cache loading and compiling into a critical section to prevent a race condition when multiple threads are compiling and caching the same code version.